### PR TITLE
Add Meilisearch endpoint

### PR DESCRIPTION
**Pull request by alfredgrip**
This pull request introduces the integration of Meilisearch into the project to enhance search capabilities. The main changes include setting up Meilisearch, configuring the environment, and implementing search-related functionality.

### Environment and Configuration:
* [`.env`](diffhunk://#diff-e9cbb0224c4a3d23a6019ba557e0cd568c1ad5e1582ff1e335fb7d99b7a1055dR68-R72): Added Meilisearch configuration variables `MEILISEARCH_HOST` and `MEILISEARCH_MASTER_KEY`.
* [`docker-compose.yml`](diffhunk://#diff-e45e45baeda1c1e73482975a664062aa56f20c03dd9d64a827aba57775bed0d3R16-R28): Added a Meilisearch service configuration.
* [`dev/setup_db.sh`](diffhunk://#diff-7d07b10677b7eac294f69762d04254cc41245e3f84e1349abec5acb71690277eL2-R2): Updated the script to use `docker compose` instead of `docker run` for setting up the database.

### Documentation:
* [`docs/reference/external-systems.md`](diffhunk://#diff-4a4d4e85f1e7fe035a9768a897218cebe1fc039b85c1a664e79007070058460fR24-R27): Documented the integration of Meilisearch and its purpose.

### Dependency Management:
* [`package.json`](diffhunk://#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519R127): Added the `meilisearch` package as a dependency.
* [`pnpm-lock.yaml`](diffhunk://#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbR109-R111): Updated to include the `meilisearch` package. [[1]](diffhunk://#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbR109-R111) [[2]](diffhunk://#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbR4088-R4090) [[3]](diffhunk://#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbR10588-R10589)

### Search Functionality:
* [`src/lib/search/meilisearch.ts`](diffhunk://#diff-0fddec15be1762b1dc5ea1350123aab2cdf73c8b1c1ca4144afacc06dde5e3b0R1-R12): Created a Meilisearch client configuration.
* [`src/lib/search/searchHelpers.ts`](diffhunk://#diff-700e58d75f9e34f06f836968a4447e134f8970475194b52993c6c709e496f3faR1-R85): Added helper functions for search operations.
* [`src/lib/search/searchTypes.ts`](diffhunk://#diff-dc751135ce243eedfc9d4a3105e06f69299532a594ee41624bb6542df3b43c8bR1-R133): Defined types and attributes for searchable entities.
* [`src/routes/(app)/api/search/+server.ts`](diffhunk://#diff-deddc71296190042ada214c4bdf812eaf1672a7c327db4a35a997a74872b253fR1-R80): Implemented a search API endpoint to handle search queries across multiple indexes.
* [`src/routes/(app)/api/search/sync/+server.ts`](diffhunk://#diff-69aef66be80e0dd3384fce63d35e272c27d08f7fe6481860a34097fee1e9c5d9R1-R260): Implemented an API endpoint to synchronize the database with Meilisearch.

### CODE REVIEW

File: src/routes/(app)/api/search/sync/+server.ts [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+      body: true,
+      bodyEn: true,
+      header: true,
+      headerEn: true,
+    },
+    where: {
+      AND: [
+        {
+          removedAt: null,
+        },
```
Comment (danieladugyan) [inline src/routes/(app)/api/search/sync/+server.ts:44]: This is not needed - our access policies already ensure this.

File: src/routes/(app)/api/search/sync/+server.ts [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+      title: true,
+      titleEn: true,
+      description: true,
+      descriptionEn: true,
+    },
+    where: {
+      AND: [
+        {
+          removedAt: null,
+        },
```
Comment (danieladugyan) [inline src/routes/(app)/api/search/sync/+server.ts:65]: This is not needed - our access policies already ensure this.

File: src/routes/(app)/api/search/meilisearch.ts [hunk 3]
```diff
@@ -0,0 +1,7 @@
+import { Meilisearch } from "meilisearch";
+import { env } from "$env/dynamic/private";
+
+export const meilisearch = new Meilisearch({
+  host: env["MEILISEARCH_HOST"],
+  apiKey: env["MEILISEARCH_MASTER_KEY"],
```
Comment (danieladugyan) [inline src/routes/(app)/api/search/meilisearch.ts:6]: We use dot access everywhere else as far as I can tell - maybe do the same here?

**Review by danieladugyan — CHANGES_REQUESTED** [review]
Some minor nitpicks but it looks awesome!

File: src/hooks.server.ts [hunk 4]
```diff
@@ -161,6 +162,12 @@ const databaseHandle: Handle = async ({ event, resolve }) => {
   return resolve(event);
 };
 
+const meilisearchHandle: Handle = async ({ event, resolve }) => {
```
[the remark this answers is not in the document]
Comment (alfredgrip) [reply src/hooks.server.ts]: Removed from locals

File: src/routes/(app)/api/search/sync/+server.ts (hunk 2 above)
Comment (alfredgrip) [reply src/routes/(app)/api/search/sync/+server.ts]: Why not double protection? :)

File: src/routes/(app)/api/search/sync/+server.ts [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+    select: {
+      id: true,
+      title: true,
+      category: true,
+      lyrics: true,
+      melody: true,
+    },
+    where: {
+      deletedAt: null,
+    },
```
[the remark this answers is not in the document]
Comment (danieladugyan) [reply src/routes/(app)/api/search/sync/+server.ts]: Hmph

File: src/routes/(app)/api/search/meilisearch.ts [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+  },
+});
+
+await Promise.all([
+  meilisearch.createIndex("positions"),
+  meilisearch.createIndex("members"),
+  meilisearch.createIndex("songs"),
+  meilisearch.createIndex("articles"),
+  meilisearch.createIndex("events"),
+]);
```
Comment (danieladugyan) [inline src/routes/(app)/api/search/meilisearch.ts:20]: Like we spotted earlier, this should not be called at the top level. It breaks CI and I guess it should only really be called once?

File: src/routes/(app)/api/search/sync/+server.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (alfredgrip) [reply src/routes/(app)/api/search/sync/+server.ts]: This access policy (and similar ones) doesn't always work as intended, ask @Macludde for more info:
`@@deny("read", removedAt != null && removedAt <= now() && !has(auth().policies, "news:article:delete"))`

[the remark this answers is not in the document]
Comment (alfredgrip) [reply src/routes/(app)/api/search/sync/+server.ts]: > This access policy (and similar ones) doesn't always work as intended, ask @Macludde for more info: `@@deny("read", removedAt != null && removedAt <= now() && !has(auth().policies, "news:article:delete"))`

Example: Go to the website, news, and search for "asdf". A deleted article will show up

**Review by danieladugyan — APPROVED** [review]

**Comment by klaratjernstrom** [comment]
LGTM!

**Pull request merged**